hugolib/page.go

```diff
@@ -369,12 +369,50 @@ func (p *pageState) TranslationKey() string {
 // AllTranslations returns all translations, including the current Page.
 func (p *pageState) AllTranslations() page.Pages {
 	p.s.h.init.translations.Do()
+
+	// Hugo attempts to reuse content providers while preparing each page for
+	// rendering. This approach means that sometimes, content providers for
+	// translations are not initialized. If needed, initialize them here.
+	for _, tr := range p.allTranslations {
+		tp, ok := tr.(*pageState)
+		if !ok {
+			continue
+		}
+		if tp.pageOutput.cp == nil {
+			cp, err := newPageContentOutput(tp, tp.pageOutput)
+			if err != nil {
+				// This will be caught by texttemplate.safeCall
+				panic(fmt.Errorf("error rendering page translation: %w", err))
+			}
+			tp.initContentProvider(cp)
+		}
+	}
+
 	return p.allTranslations
 }
 
 // Translations returns the translations excluding the current Page.
 func (p *pageState) Translations() page.Pages {
 	p.s.h.init.translations.Do()
+
+	// Hugo attempts to reuse content providers while preparing each page for
+	// rendering. This approach means that sometimes, content providers for
+	// translations are not initialized. If needed, initialize them here.
+	for _, tr := range p.translations {
+		tp, ok := tr.(*pageState)
+		if !ok {
+			continue
+		}
+		if tp.pageOutput.cp == nil {
```

Comment:
Thanks for this.

But I don't think we can solve this like this. Most notably is the data race which, if it's not already failing the build, it will if you move the `.Translations` into the single template.

I suspect that the cleanest, working solution we will find without doing too much work is to expand on the patch we added a few days ago: Make sure that the lazy content provider is triggered on all access to `cp` (including `RenderString`).

Reply:
Thanks for the feedback. Would you mind explaining the data race a bit more? Which goroutines, processes, I/O, etc. would be racing?

I'll spend some time early this week implementing the approach you've suggested.

Reply:
1. We render pages in parallel.
2. Given the below, `AllTranslations` may be called from multiple go routines.
3. Any of the code lines reading or writing mutable page state below will be a potential data race (e.g. `if tp.pageOutput.cp == nil` and `tp.initContentProvider(cp)`). Go's race detector will fail the tests when run with `go test -race` (which we do on the CI server), but I guess you need to have test cases that trigger it.

```go
func (p *pageState) AllTranslations() page.Pages {
	p.s.h.init.translations.Do()

	// Hugo attempts to reuse content providers while preparing each page for
	// rendering. This approach means that sometimes, content providers for
	// translations are not initialized. If needed, initialize them here.
	for _, tr := range p.allTranslations {
		tp, ok := tr.(*pageState)
		if !ok {
			continue
		}
		if tp.pageOutput.cp == nil {
			cp, err := newPageContentOutput(tp, tp.pageOutput)
			if err != nil {
				// This will be caught by texttemplate.safeCall
				panic(fmt.Errorf("error rendering page translation: %w", err))
			}
			tp.initContentProvider(cp)
		}
	}

	return p.allTranslations
}
```